src/generateContainer.js

```diff
@@ -1,69 +1,98 @@
+    const { routeClass, params: getParams = defaultGetParams} = currentNode;
+
+    // Not using XOR here allows better invariant error messages.
+    invariant(
+      route || routeClass,
+      `relay-nested-routes: Route with component \`${componentName}\` is `+
+      `missing a route or routeClass prop: ` +
+      `<Route component={${componentName}} route={...}/> or` +
+      `<Route component={${componentName}} routeClass={...}/> or`
+    );
+    invariant(
+      !(route && routeClass),
+      `relay-nested-routes: Route with component \`${componentName}\` ` +
+      `specifies both route and routeClass`
+    );
+
+    const routeParams = getParams({
+      params: newProps.params,
+      query: newProps.location.query
+    });
+    Object.keys(routeParams).forEach(paramKey => {
+      if (params[paramKey !== undefined]) {
+        invariant(
+          params[paramKey] === routeParams[paramKey],
+          `relay-nested-routes: Route with component \`${componentName}\`` +
+          ` has conflict on param \`${paramKey}\``
+        );
+      }
+
+      params[paramKey] = routeParams[paramKey];
+    });
+
+    // Explicitly constructing the route from the class allows us to re-use
+    // any invariants like required params on the route constructor.
+    if (routeClass) {
+      route = new routeClass(params);
+    }
+
+    const routeName = route.name;
+    allRouteNames.push(routeName);
+
     const fragmentResolvers = [];
 
-    if (Relay.isContainer(Component)) {
-      const { route } = branch[index];
-      invariant(
-        route,
-        `relay-nested-routes: Route with component ` +
-        `\`${Component.displayName}\` is missing a route prop: ` +
-        `<Route component={${Component.displayName}} route={...}/>`
-      );
-
-      Object.keys(route.queries).forEach(queryName => {
-        const newQueryName = `Nested_${route.name}_${queryName}_${++queryIdx}`;
-        fragments[newQueryName] = queries[newQueryName] = (_, ...args) => {
-          return route.queries[queryName](Component, ...args);
-        };
-        fragmentResolvers.push({
-          prop: queryName,
-          resolve: function getLocalProp() {
-            return this.props[newQueryName];
-          }
-        });
+    Object.keys(route.queries).forEach(queryName => {
+      const newQueryName = `Nested_${routeName}_${queryName}_${++queryIdx}`;
+      fragments[newQueryName] = queries[newQueryName] = (_, ...args) => {
+        return route.queries[queryName](Component, ...args);
+      };
+      fragmentResolvers.push({
+        prop: queryName,
+        resolve: function getLocalProp() {
+          return this.props[newQueryName];
+        }
       });
-    }
+    });
 
     return function ComponentGenerator(props) {
       fragmentResolvers.forEach(fragment => {
         props[fragment.prop] = fragment.resolve.call(this);
       });
 
-      return <Component {...props}/>;
+      return <Component {...props} {...routeParams}/>;
```

Comment:
This is what `RootContainer` does to inject `params`, per https://github.com/facebook/relay/blob/v0.1.1/src/container/RelayRootContainer.js#L295. It's necessary for the container to resolve the params as variables correctly, such that it can pull the correct data from the query.

Not sure if we need extra documentation here - upstream doesn't have any, and it took me a while to figure out what was going on... but nobody else should need to depend on this behavior (hopefully).